branchpro/new_posterior.py

```diff
@@ -0,0 +1,1521 @@
+                tau_window = (tau_window[1:] +  # noqa
+                              [self._infectious_individuals(self.cases_data,
+                                                            end_window-1)])
+
+            except UnboundLocalError:
+                # First iteration, so set up the sliding window
+                tau_window = self._infectives_in_tau(
+                    self.cases_data, start_window, end_window)
+
+            Ll += np.log(r_profile[_]) * np.sum(slice_cases)
+
+            log_tau_window = np.zeros_like(tau_window)
+            for tv, tau_val in enumerate(tau_window):
+                if tau_val == 0:
+                    log_tau_window[tv] = 0
+                else:
+                    log_tau_window[tv] = np.log(tau_window[tv])
+
+            Ll += np.sum(np.multiply(slice_cases, log_tau_window))
+            Ll += - r_profile[_] * np.sum(tau_window)
+
+        return Ll
+
+    def _compute_derivative_log_likelihood(self, r_profile):
+        """
+        """
+        total_time = self.cases_times.max() - self.cases_times.min() + 1
+        time_init_inf_r = self._tau + 1
+
+        dLl = []
+
+        for _, time in enumerate(range(time_init_inf_r+1, total_time+1)):
+            # get cases in tau window
+            start_window = time - self._tau
+            end_window = time + 1
+
+            try:
+                # try to shift the window by 1 time point
+                tau_window = (tau_window[1:] +  # noqa
+                              [self._infectious_individuals(self.cases_data,
+                                                            end_window-1)])
+            except UnboundLocalError:
+                # First iteration, so set up the sliding window
+                tau_window = self._infectives_in_tau(
+                    self.cases_data, start_window, end_window)
+
+            dLl.append((1/r_profile[_]) * np.sum(
+                    self.cases_data[(start_window-1):(end_window-1)])
+                    - sum(tau_window))
+
+        return dLl
+
+    def evaluateS1(self, x):
+        """
+        """
+        # Compute log-likelihood
+        try:
+            Ll = self._compute_log_likelihood(x)
+
+            # Compute derivatives of the log-likelihood
+            dLl = self._compute_derivative_log_likelihood(x)
+
+            return Ll, dLl
+
+        except ValueError:  # pragma: no cover
+            return -np.inf, [-np.inf] * self.n_parameters()
+
+    def __call__(self, x):
+        """
+        Evaluates the log-likelihood in a PINTS framework.
+
+        Parameters
+        ----------
+        x : list
+            List of free parameters used for computing the log-likelihood.
+
+        Returns
+        -------
+        float
+            Value of the log-likelihood at the given point in the free
+            parameter space.
+
+        """
+        try:
+            return self._compute_log_likelihood(x)
+
+        except ValueError:  # pragma: no cover
```

Comment:
These are very useful but I am also wary of them in case they hide a bug in the computation of the log likelihood.
What about also doing a RuntimeWarning if this happens saying what x was and that the likelihood was returned as -infinity.